R/tags.R

```diff
@@ -288,6 +295,97 @@ tagFunction <- function(func) {
   structure(func, class = "shiny.tag.function")
 }
 
+#' Add a tag render function
+#'
+#' Adds a render method to a tag object. This allows for the tag structure to
+#' be physically present while still allowing for a function to enhance (or
+#' even completely replace) the original tag object.
+#'
+#' It is recommended to use a render function over a [`tagFunction()`] whenever
+#' possible. By using a render method, a the tag structure is not a black box
+#' and can be inspected and altered before print time.
+#'
+#' Using [`tagFunction()`] is recommended if a stand-in tag structure does not
+#' make sense.
+#'
+#' @seealso [`tagFunction`]
+#' @param tag A [`tag()`] object.
+#' @param func Function with at least one argument (the `tag`).
+#' @param add If `TRUE`, the previous render function is called before calling
+#'   this `func`. Otherwise, any previous render function is ignored.
+#' @return A [`tag()`] object with a `.render` field containing `func`.
+#'   When the returned tag is _rendered_ (such as with [`as.tags()`]),
+#'   this function will be called.
+#' @examples
+#' # Have a place holder div and return a span instead
+#' obj <- div("example", .render = function(x) {
+#'   x$name <- "span"
+#'   x
+#' })
+#' obj$name # "div"
+#' print(obj) # Prints as a `span`
+#'
+#' # Add a class to the tag
+#' # Should print a `span` with class `"extra"`
+#' spanExtra <- tagAddRender(obj, function(x) {
+#'   tagAppendAttributes(x, class = "extra")
+#' })
+#' spanExtra
+#'
+#' # Replace the previous render method
+#' # Should print a `div` with class `"extra"`
+#' divExtra <- tagAddRender(obj, replace = TRUE, function(x) {
+#'   tagAppendAttributes(x, class = "extra")
+#' })
+#' divExtra
+#'
+#' # Add more child tags
+#' spanExtended <- tagAddRender(obj, function(x) {
+#'   tagAppendChildren(x, " ", tags$strong("bold text"))
+#' })
+#' spanExtended
+#'
+#' # Add a new html dependency
+#' newDep <- tagAddRender(obj, function(x) {
+#'   fa <- htmlDependency(
+#'     "font-awesome", "4.5.0", c(href="shared/font-awesome"),
+#'     stylesheet = "css/font-awesome.min.css")
+#'   attachDependencies(x, fa, append = TRUE)
+#' })
+#' # Also add a jqueryui html dependency
+#' htmlDependencies(newDep) <- htmlDependency(
+#'   "jqueryui", "1.11.4", c(href="shared/jqueryui"),
+#'   script = "jquery-ui.min.js")
+#' # At render time, both dependencies will be found
+#' renderTags(newDep)$dependencies
+#'
+#' # Ignore the original tag and return something completely new.
+#' newObj <- tagAddRender(obj, function(x) {
+#'   tags$p("Something else")
+#' })
+#' newObj
+tagAddRender <- function(tag, func, add = TRUE) {
```

Comment:
Thinking about this again, maybe this could be `tagRenderHook()` and `.render` could be `.renderHook` (I was wrong about htmlwidgets using `onRender` -- it's actually `preRenderHook`)?

Reply:
Works for me!